src/Compilers/CSharp/Test/Semantic/Semantics/BindingAsyncTests.cs

```diff
@@ -1233,10 +1282,113 @@ class A
+                // error CS5001: Program does not contain a static 'Main' method suitable for an entry point
+                Diagnostic(ErrorCode.ERR_NoEntryPoint).WithLocation(1, 1));
+            var entry = compilation.GetEntryPoint(CancellationToken.None);
+            Assert.Null(entry);
+        }
+
+        [Fact]
+        public void MainCanBeAsyncAndGenericOnIntWithArgs()
+        {
+            var origSource = @"
+using System.Threading.Tasks;
+
+class A
+{
+    async static Task<int> Main(string[] args)
+    {
+        return await Task.Factory.StartNew(() => 5);
+    }
+}";
+            var sources = new string[] { origSource, origSource.Replace("async ", "").Replace("await", "") };
+            foreach (var source in sources)
+            {
+                var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugExe);
+                compilation.VerifyDiagnostics();
+                var entry = compilation.GetEntryPoint(CancellationToken.None);
+                Assert.NotNull(entry);
+                Assert.Equal("System.Threading.Tasks.Task<System.Int32> A.Main(System.String[] args)", entry.ToTestDisplayString());
+            }
+        }
+
+        [Fact]
+        public void MainCanBeAsyncAndGenericOnInt()
+        {
+            var origSource = @"
+using System.Threading.Tasks;
+
+class A
+{
+    async static Task<int> Main()
+    {
+        return await Task.Factory.StartNew(() => 5);
+    }
+}";
+            var sources = new string[] { origSource, origSource.Replace("async ", "").Replace("await", "") };
+            foreach (var source in sources)
+            {
+                var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugExe);
+                compilation.VerifyDiagnostics();
+                var entry = compilation.GetEntryPoint(CancellationToken.None);
+                Assert.NotNull(entry);
+                Assert.Equal("System.Threading.Tasks.Task<System.Int32> A.Main()", entry.ToTestDisplayString());
+            }
+        }
+
+        [Fact]
+        public void MainCantBeAsyncAndGenericOverFloats()
+        {
+            var source = @"
+using System.Threading.Tasks;
+
+class A
+{
+    async static Task<float> Main()
+    {
+        await Task.Factory.StartNew(() => { });
+        return 0;
+    }
+}";
+            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.ReleaseExe);
+            compilation.VerifyDiagnostics(
+                // (6,30): warning CS0028: 'A.Main()' has the wrong signature to be an entry point
+                //     async static Task<float> Main()
+                Diagnostic(ErrorCode.WRN_InvalidMainSig, "Main").WithArguments("A.Main()").WithLocation(6, 30),
+                // error CS5001: Program does not contain a static 'Main' method suitable for an entry point
+                Diagnostic(ErrorCode.ERR_NoEntryPoint).WithLocation(1, 1) );
         }
```

Comment:
Never mind, I see that you use string replacement to achieve that.

---
In reply to: [108235403](https://github.com/dotnet/roslyn/pull/18157#discussion_r108235403) [](ancestors = 108235403)